### ISSUE TITLE
Using openMSX in build pipeline

### ISSUE THREAD

**Issue by immetoo2** [openMSX/openMSX#1133]
Hi, I'm tring to use openMSX as one or more build steps to package msx software.
For example to run the orginal file packers,custom utils and unittests.

Its working great already when manually closing openMSX constantly.
It would be nice to automate this, for this I'm requesting the following features;

* Add CommandDevice to send command from msx to emulator.
    Example usecases;
    * omsxctl exit
    * omsxctl set speed 999
    * omsxctl plug joyporta mouse
    * omsxctl set fail_after 100000

* Add optional exitCode argument to exit command.
    So msx can do shutdown with error to flag build process.

* Add failAfter command + command line option.
    When starting openMSX with this command line option it
    will always do exit(1) after the timeout.
    Expect for when the normal exit command is executed.
    It should be allowed to change this value runtime.

Is this something which is able to be added to openMSX ?

Have created a ~working commanddevice will add a PR for review.
The other two features I not sure how and where to impl it best.

**Comment by MBilderbeek** [comment]
Just to understand better what you're attempting to do here, I have some questions to trigger some discussion.

Wouldn't this also be possible to implement with only a Tcl script? E.g. you could write a script that puts an I/O watchpoint on a certain unused Z80 port and the MSX program writes data to that port, which the scripts interprets.

For the third thing: there is already quite some timing possible in Tcl. A Tcl script could simply check some condition after a certain amount of MSX time and then perform the exit (with or without flag).

For the 2nd thing: so, you basically want some simple and easy way to communicate back to the program that started openMSX. I suppose the CliComm could be used for this as well, but it's not as easy as this. CliComm allows you to give commands to openMSX remotely. Then the remote program is running these Tcl commands to check certain conditions and establish a 'fail' or 'success' state of the running MSX program, see http://openmsx.org/manual/openmsx-control.html

Anyway, as I said, just asking some questions to trigger discussion, to see what the best way would be to solve the use cases you're proposing, which do sound quite interesting!

Also wondering what the other devs think.

**Comment by m9710797** [comment]
I like this idea.

Let's first answer your last two points.

* Extend the openMSX 'exit' command with an optional exitcode. Shouldn't be
too hard to implement. I'll try to add it.
Manuel's suggestion (use CliComm for communication between openMSX and the
other process) would technically work, but it's too hard for a simple use
case like using openMSX as a tool in a build pipeline.

* fail_after can be implemented like this
      proc fail_after {time} { after $time {exit 1} }
And because it's this simple maybe it doesn't even need a helper proc.
If needed this can be extended so that the timeout value can be changed (by
canceling the current after-event and scheduling a new one).
I don't understand why you need an openMSX command line option for this.


* Now the main point, the communication mechanism between the MSX program
and openMSX.

1) Should this be implemented as a MSXDevice (like in the current
prototype) or as a Tcl script. Technically both are possible.

We already have the 'debugdevice', that is implemented as a MSXDevice. So
it makes sense to implement 'commanddevice' in a similar way.
We also have Tcl scripts that use the IO-watchpoint approach, though
usually those watch existing MSXDevices (e.g. the vgmrecorder script). I
don't think we have any scripts that use IO-watchpoints to create
completely new devices.

In openMSX we have this goal:
   "A MSX program should not be able to detect whether it's running on a
real MSX or on an emulator."
Therefor the 'omsxctl'-mechanism must be opt-in.

In the approach of a MSXDevice, the user must pass the '-ext commanddevice'
option to openMSX, that's clearly opt-in. The same opt-in can be achieved
by a script that's not installed in openmsx/share/scripts but must be
activated by passing '-script omsxctl.tcl' on the command line.

Detail: currently the extension is called 'commanddevice' and the MSX-DOS
tool that uses this device is called 'omsxctl' (because of the 8-char
limit). Maybe it would be less confusing to pick the same name for both?


2) What is the actual mechanism used for the communication?

It tried to reverse-engineer this from the c++ code. Is the following a
fair summary?

- 'commanddevice' is a 'switched device', meaning it uses the IO area
0x40-0x4F (this area is possibly shared between multiple devices).
- To activate 'commandevice' write the value 30 to IO port 0x40. When
active, reading port 0x40 returns the value 225 (30 xor 255), this can be
used to detect the presence of this device. (I like this detection
possibility).
- To send a command string to openMSX, write the sequence of characters to
port 0x41 (this fills the send-buffer). (Detail: the string is UTF8
encoded, normally this doesn't matter, but be careful with chars >= 128).
When the full string is passed to the send-buffer, write any value to port
0x42. Now openMSX will execute the command. Executing a command clears the
send-buffer. If for some reason you want to clear the buffer without
executing the command, then write (any value) to port 0x43.
- Once a command has executed, you can read the status of that execution
from port 0x41 (0=success, 1=error). The result of the command (result
string or error message) can be read character-per-character from port
0x42. If the last character has been read, the status (port 0x41) changes
to value 255. (Thus before reading each character from 0x42, you should
check port 0x41.) The result string does get transformed before being
passed (from Tcl) to the commanddevice: '\n' characters (10) are replaced
by '\r\n' (13 10), this way the string can directly be printed by e.g.
MSX-DOS. If for some reason you want to skip reading part of the
result-string or if you want to re-read from an earlier position, you can
write a position index to port 0x44.

At least this is how I understood the code. Is it correct?

I do see a few problems in this protocol:
- Suppose the command does return an empty string. Then reading port 0x41
immediately returns 255, and I cannot test whether the command executed
successfully or with an error.
- I don't like the automatic translation of \n into \r\n. Usually the
result of a Tcl command is a string, and then this translation is useful.
But it could also be binary data (e.g. generated by the Tcl command 'binary
format ....') and in that case this transformation will damage the result.
This translation is easy to do in the 'omsxctl' application, so I propose
to move it there.
- Is the cancel-a-partially-send-command (port 0x43) a useful feature? The
'omsxctl' application doesn't use it. I also can't think of any use case
for it.
- Is the seek-in-return-string a useful feature? Maybe it is, but i don't
see it right now. (The 'omsxctl' application doesn't use it). Also the seek
index is limited to 8-bit, while in principle the result could be longer
than 256 characters.

I'm unsure about "write anything to port 0x42 to signal the end of the
command". The alternative I'm considering is "write a zero-terminated
string to 0x41". In other words: the command starts executing after a zero
is written to 0x41. Note that commands are always plain strings and cannot
legally contain zero-characters (on the other hand the result-"string"
could in principle be binary data which could contain embedded zeroes). The
(only) advantage of this alternative is that it may result in simpler MSX
code (e.g. a command can then be send using a single OTIR Z80 instruction).

Maybe the status port (0x41) should be split in two independent bits. One
bit to signal "there are still unread characters in the result". And
another bit to indicate "last command was successful or error".


3) Can/should we generalize the communication mechanism to other use cases?

The above communication mechanism (with or without my suggested changes) is
well suited for the original use case (using openMSX as a tool in a build
pipeline). For this use case it doesn't really matter how long it takes to
send the command or to receive the result (the Z80 writes/reads the
command/result character-by-character). An alternative communication
mechanism could be:
  LD HL,CommandString  ; pointer to zero-terminated (utf8) string
  LD DE,ResultBuffer        ; the result will be written here
  LD BC,ResultBufSize     ; maximum length for the result (avoid buffer
overflow) could be 0 if you're not interested in the result (and then the
value of DE doesn't matter)
  OUT (#2D),A                   ; (value of A doesn't matter)
  JP C,Error                       ; carry-flag set on error
The write to port 0x41 is a trigger that invokes "emulator magic": openMSX
will read the content of the HL,BC,DE registers, read the string from
memory, execute the command, write the result back to memory, and update
the carry flag. (And maybe the BC register could be used to indicate the
length of the result (and the value 0xffff could mean that the given result
buffer was too small to hold the complete result)).
This whole operation runs in zero-elapsed-msx-time. And this may (or may
not) be a useful property for a particular use case.

Now that I think about, this alternative is easily implemented purely as a
Tcl script with an IO-watchpoint (like Manuel suggested). Maybe I'll try
this, to be able to more easily compare both alternatives.

I'm not sure yet which alternative I like best. Opinions?

Wouter

**Comment by m9710797** [comment]
I've created the script I talked about in the previous comment. See attachment: [omsxctl.tcl.txt](https://github.com/openMSX/openMSX/files/2695837/omsxctl.tcl.txt).  See comments in the script for an example of how to use it. I only did a few tests so far, but it seems to work quite well.

Next I'll see if I can get "exit [&lt;exitcode&gt;]" to work.

**Comment by m9710797** [comment]
Ok, exit now takes an optional exit-code parameter.
I also tested that stuff like `after 5000 exit 9` works.

**Comment by m9710797** [comment]
Ok, I pushed an implementation of this idea (implemented via the Tcl script approach), including a modified version of omsxctl.com.

@immetoo: Please check that this version still covers all your use cases.

Some small remarks:
- Disabling automatic setting saving can be done with `set save_settings_on_exit false`.
- Instead of `set speed 999` you might want to use `set throttle off` (instead of run 9.99x faster, run as fast as possible).
- Passing information from the host shell to openMSX scripts could be done via environment variables (instead of extending openMSX with extra command line options). Reading env vars in Tcl can be done via `$::env(MY_ENV_VAR)`.

The 'after_fail' functionality could maybe implemented like this?
`  omsxctl set id [after time 10 "exit 1"]`
`   makelib1`
`   omsxctl after cancel id`
`   omsxctl set id [after time 20 "exit 1"]`
`   makelib2`
`   omsxctl after cancel id`
`   omsxctl exit 0`

**Comment by immetoo2** [comment]
Have been playing with it, it is a lot faster with the new tcl impl.
Have created my first tcl scripts to complete the build helper scripts;
[boot_fast.tcl.txt](https://github.com/openMSX/openMSX/files/2704531/boot_fast.tcl.txt)
[boot_stateless.tcl.txt](https://github.com/openMSX/openMSX/files/2704532/boot_stateless.tcl.txt)
[boot_timeout.tcl.txt](https://github.com/openMSX/openMSX/files/2704533/boot_timeout.tcl.txt)
[fail_after.tcl.txt](https://github.com/openMSX/openMSX/files/2704534/fail_after.tcl.txt)

I could not get the 'after cancel' working but could work around it.
Now have this [autoexec.bat.txt](https://github.com/openMSX/openMSX/files/2704549/autoexec.bat.txt) working and it outputs;
=== Boot watchdog timeout
25
=== Disable boot watchdog
off
=== Test fail_after
Scheduled automatic failure in: 3600
Scheduled automatic failure in: 100 and stopped previous attempt.
Scheduled automatic failure in: 50 and stopped previous attempt.
Scheduled automatic failure in: 3 and stopped previous attempt.
=== below ~3 empty echo is too slow :)
Disabled automatic failure and stopped previous attempt.
=== Plug mouse
OK
=== Set speed
350
=== Enable speed throttling
true
=== Start program
C>

So that is working great, still have some questions;
* What is the time unit ? it scaled with speed and works oke...
  But maybe using `normal time`(ms) is also possible ?
* omsxctl.com outputs looks like it needs an extra line ending, see all extra echos
* omsxctl.z80:10 "exit 1" should have quotes
* Is it possible to include these tcl scripts and com in build for make install so we don't need to copy it in 
  every repro ? Besides my wip projects have spotted multiple repros from @fr3nd which could use
  these build integration helpers as well.
* offtopic: the plug command has no success message

On some later date i want to package openMSX in an docker image to get real public ci pipeline working but that can only work on roms.
But it would be nice if we can boot a more working msx2 type system are there efforts besides c-bios which impl the other roms so nextor can boot ?

**Comment by m9710797** [comment]
> I could not get the 'after cancel' working ...

This version ([fail_after.tcl.txt](https://github.com/openMSX/openMSX/files/2704985/fail_after.tcl.txt)) seems to work for me.

> What is the time unit ? it scaled with speed and works oke...
But maybe using normal time(ms) is also possible ?

Try `help after` in the openMSX console. Time unit is seconds, but it accepts floating point values (so pass `[expr {$my_time / 1000.0}]` if `my_time` is in milliseconds. `after time` uses MSX time (depends on `speed` and `throttle` settings), `after realtime` uses real time. I assume that for timeouts MSX time makes more sense.

> omsxctl.com outputs looks like it needs an extra line ending, see all extra echos
> omsxctl.z80:10 "exit 1" should have quotes

Thanks. Both fixed now.

> Is it possible to include these tcl scripts and com in build for make install ...

What do you mean exactly? Where should it be installed? I don't believe we install any other stuff from Contrib/. I explicitly don't want to have `omsxctl.tcl` installed in the `openmsx/share/scripts` directory.

> offtopic: the plug command has no success message

This is indeed not always consistent in openMSX. My personal preference is that commands are silent on success.

**Comment by immetoo2** [comment]
My bad i think i tried that one also but must been late, have now removed the workaround.
Also moved to realtime as explaining seconds is easy relative to an floating time unit.
And added for consistency with fail_afer also `set boot_wdt 0` support;
[boot_timeout.tcl.txt](https://github.com/openMSX/openMSX/files/2705250/boot_timeout.tcl.txt)
[fail_after.tcl.txt](https://github.com/openMSX/openMSX/files/2705251/fail_after.tcl.txt)

Can we put the 4 extra tcl script in Contrib/cibuild ?
And about contrib its fine that is not in `openmsx/share/scripts` but having it in ~`openmsx/share/contrib` (or even a single tar/zip) would be nice for all user contributions which are now not accessible from the distribution.
In build scripts side for all tools we have single paths like `PATH_OPENMSX ?= /opt/openMSX`.
From there we can copy/use/activate the user contribs when needed.
imho, the workaround to have `PATH_OPENMSX_BIN` and `PATH_OPENMSX_SRC` as an external dependency when using any of the contributions is not nice to the end user.

**Comment by MBilderbeek** [comment]
Just curious: why would you use realtime and not MSX time? The MSX-time should be constant, but the realtime depends on all kinds of external situations... (PC load, video card, speed settings, frameskip settings, etc.)
Also MSX time can be used in floating point.

**Comment by immetoo2** [comment]
While playing with popcom got a nice failure case for which the boot watchdog was intented for;

`BOOT_TIMEOUT=25;/opt/openMSX/bin/openmsx .... -diska bin/dist-run`
`warning: Virtual diskimage full: menu_ext.ge5 truncated.`
`warning: Virtual diskimage full: command2.com truncated.`
`make: *** [src/dist-run/0module.mk:25: dist-run] Error 124`

`real    0m25.239s`
`user    0m24.446s`
`sys     0m1.057s`

When using msx time the value could be
`real    0m1.443s`
while we requested 25.

Same goes for writing automation or tests with fail_after with for example 120 seconds timeout as developer i run it on different speeds like build=max, build-debug=1x, test=custom, and seeing the timeouts using different values as written is always an extra think step like oh yeah its emu time.
But having the option to have both is also nice like;
omsxctl fail_after 120 msxtime
Would raise no thoughts seeing it fail in 2 seconds instead of 120 on fast machine.

**Comment by MBilderbeek** [comment]
Well, if you want to check on whether something on the MSX is taking not too long (e.g. the execution of something running on the MSX, like a program, etc.), you need normal 'time', but if you want to check whether your total build is not taking too long (due to some error interrupting it, as you showed above), you probably want to use 'realtime'. But such a generic sanity-check time-out I would implement outside openMSX myself.

**Comment by m9710797** [comment]
Hi,

I'm not sure I like this approach. You try to be very generic in these scripts, but at the same time the scripts are also not generic enough.

This seems to promote a style like:
`    FAIL_AFTER_BOOT=30 \`
`    FAIL_AFTER_PATH=bin/myapp \`
`    BOOT_HDD_PATH=bin/myapp/disk \`
`    BOOT_HDD_IMAGE=bin/myapp/disk.img \`
`    RENDERER=none \`
`    THROTTLE=off \`
`    openmsx \`
`      -script <path-to>/boot_env.tcl \`
`      -script <path-to>/boot_hdd.tcl \`
`      -script <path-to>/fail_after.tcl \`
`      -machine ...`
Instead I prefer a style more like (I did not try to replicate the exact same functionality in both examples):
`    echo \`
`      set renderer none \;\`
`      set throlle off \;\`
`      diskmanipulator create disk.img 32m 32m 32m 32m \;\`
`      hda disk.img \;\`
`      diskmanipulator import hda1 ./disk-in/ \;\`
`      after quit {diskmanipulator export hda1 ./disk-out/} \;\`
`    > start.tcl`
`    openmsx -script start.tcl`

* The first style is more abstract. The second is more direct.
* The first style is easier if the pre-written scripts (boot_env.tcl, ...) already provide exactly what you want. The second style is more flexible because you can write any command with whatever options you need.
* The second style requires the user to know about openMSX commands. In the first style that's not needed. OTOH the first style requires the user to know about the specific environment variables. The openMSX commands should already be documented. These enviroment variables require additional documentation.

This feature is meant for technically capable users. So IMHO we should promote the second style which is more flexible but maybe has a little steeper learning curve.

Maybe these scripts can be reworked into a text document which has lots of examples. Users can then copy/paste parts of the examples into their own build-flow.

In other words I'd like to promote a style which encourages users to try new things themselves and come up with new ideas, rather than promote a style which is limited to the things only we thought about.


I do see some value in the fail_after.tcl script. Though in practice I think starting openMSX like
`/usr/bin/timeout 20 openmsx ...`
is often already good enough. The automatic screenshot functionality is a nice bonus, but is the worth the extra complexity?

Instead of the harddisk import/export I usually use the DirAsDisk functionality. Though of course that's (still) limited to 720kB.

**Comment by immetoo2** [comment]
Indeed this is for power users, but in no way i try to promote a special
boot/config/script style to anyone or to exclude or dispromote having additional tcl scripts doing more things.

This main part of this contribution are three example tcl scripts from
which can be used and learned from and be modified to own usage.
Which cover most pitfalls about error handling and exiting the process.

As they are in contrib the readme states;
'These packages are not part of openMSX'
This means they can't encourages or promote anything for openMSX.
So are these scripts not just like other contributions for example 'basictorom.tcl' script ?

The DirAsDisk i used it a lot and works great but import/export via hdd is
a lot faster. (9s vs 16s running 12x)

**Comment by MBilderbeek** [comment]
The latter is probably due to disk drive speed emulation.

**Comment by immetoo2** [comment]
Closing as omsxctl mechanism is sufficient, will move it all to new repro and use that.
